ocaml/parsing/ast_invariants.ml

```diff
@@ -259,7 +259,10 @@ let iterator =
        checks for attribute payloads.  See comment on [attr_tracking_time] in
        [builtin_attributes.mli]. *)
     super.attribute { self with attribute = super.attribute } attr;
-    Builtin_attributes.(register_attr Invariant_check attr.attr_name)
+    (* CR zqian: I'm not sure about this dirty hack. Basically, mode-syntax
+       attributes should not be registered - how to do that cleanly? *)
```

Comment:
Can we just register them and mark them as used when we read them in `typemexp.ml`? (Or, if it's difficult to do it there, we could mark them as used in Jane Syntax in `partition_attributes` and the like.) I would prefer not to make the logic here more complicated.

Reply:
Ideally they shouldn't be registered - because they are always parsed at proper places which is guaranteed to be used. 
But I agree that this bit here is just as ugly.

(We can't mark them as used in `Jane_syntax` because that would cause module-dependency-cycle)

I will mark it in `typemexp.ml`.

Reply:
In the end I think it's probably least evil to keep it as it is. 

To do it properly, we should register attributes in the parser, and mark them as used in `Jane_syntax.Mode_expr.of_attrs`. However, `Builtin_attributes` depends on `Jane_syntax` so do that would cause a cycle.

We can't mark them in `typemexp` - it doesn't know anything about attributes. In particular, `register_attr` requires the location of attribute name, which `typemexp` doesn't know about - it has the location of the mode expression, but not the attribute name. 

So we should mark them in `typecore`. But would ruin the interface of `Jane_syntax.Mode_expr.attr_of` - it would need to return the attribute name with location. I would rather not do that.

Reply:
This is solved by commit "add mode syntax to ast_iterator".